R/GeospatialFunctions.R

```diff
@@ -0,0 +1,571 @@
+                      sf::st_as_sf(coords = c("lon", "lat"),
+                                   crs = unique(.x$epsg)) %>%
+                      # transform to the selected CRS:
+                      sf::st_transform(sf::st_crs(as.numeric(crs)))) %>%
+      dplyr::select(-epsg)
+  }))
+  
+  return(sf)
+  
+}
+
+
+#' fetchATTAINS
+#' 
+#' Fetch ATTAINS features within a bounding box produced from a set of TADA spatial features.
+#' 
+#' @param data A dataframe developed using `TADA_DataRetrieval()` or `TADA_MakeSpatial()`.
+#' @param type The type of ATTAINS data you would like to fetch ("lines", "points", "polygons", "catchments"). All ATTAINS features are returned in WGS84 (crs = 4326).
+#' 
+#' @return spatial features that are within the spatial bounding box of water quality observations.
+#' 
+#' @seealso [TADA_MakeSpatial()]
+#' @seealso [TADA_DataRetrieval()]
+#' 
+#' @examples
+#' \dontrun{
+#'tada_data <- TADA_DataRetrieval(startDate = "1990-01-01",
+#'                                endDate = "1995-12-31",
+#'                                characteristicName = "pH",
+#'                                statecode = "NV",
+#'                                applyautoclean = TRUE)
+#'                                  
+#'nv_attains_lines <- fetchATTAINS(data = tada_data, type = "lines")
+#' }
+
+fetchATTAINS <- function(data, type = NULL) {
+  
+  if(is.null(data) | nrow(data) == 0){
+    stop("There is no data in your `data` object to use as a bounding box for selecting ATTAINS features.")
+  }
+  
+  # If data is already spatial, just make sure it is in the right CRS
+  # and add an index as the WQP observations' unique IDs...
+  if (!is.null(data) & inherits(data, "sf")) {
+    data <- data %>%
+      sf::st_transform(4326) 
+  } else {
+    # ... Otherwise transform into a spatial object then do the same thing:
+    data <- data %>%
+      #convert dataframe to a spatial object
+      TADA_MakeSpatial(data = ., crs = 4326) 
+  }
+  
+  if(is.null(type)){
+    stop("Please select the type of ATTAINS data you would like to import ('catchments', 'lines', 'points', or 'polygons').")
+  }
+  
+  # select baseurl based on type of ATTAINS data you want to download:
+  if(type == "catchments"){
+    baseurl <- "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/3/query?"
+  }else if(type == "points"){
+    baseurl <-  "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/0/query?"
+  }else if(type == "lines"){
+    baseurl <-  "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/1/query?"
+  }else if(type == "polygons"){
+    baseurl <-  "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/2/query?"
+  }
+  
+  #starting at feature 1 (i.e., no offset):
+  offset <- 0 
+  # empty list to store all features in
+  all_features <- list()
+  
+  # EPSG we want our ATTAINS data to be in (always 4326 for this function)
+  epsg <- sf::st_crs(data)$epsg
```

Comment:
Since `epsg` should always be 4326 it seems like you could remove the definition here and save a couple lines. 

I was curious about `offset` for a similar reason but I'm realizing I might be wrong about that

Reply:
agree on setting epsg <- 4326! 

offset needs to be explicitly set to 0 at first, then changes with each "repeat"